### PR TITLE
MULE-18993: Generate an errorTypeRepository from the ArtifactAst

### PR DESCRIPTION
**Pull request by elrodro83**

### CODE REVIEW

**Comment by elrodro83** [comment]
--validate

**Comment by elrodro83** [comment]
--validate

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java [hunk 1]
```diff
@@ -630,9 +630,11 @@ public void appliesPolicyUsingErrorTypeAndHavingDependencies() throws Exception
   @Test
   @Issue("MULE-18196")
   @Description("The application declares an ErrorType that is needed by the policy and but the policy doesn't have it in its own ErrorType repository")
-  public void appliesPolicyUsingErrorTypeDeclaredOnApp() throws Exception {
+  public void appliesPolicyUsingErrorTypeDeclaredOnAppDependency() throws Exception {
```
Comment (szaffarano) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java:633]: What about `MULE-18196`? Does this failing test mean that we introduced a regression?

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/error/ErrorTypeBuilder.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  private boolean isOrphan() {
+    return (identifier.equals(ANY_IDENTIFIER) || identifier.equals(CRITICAL_IDENTIFIER)) && namespace.equals(CORE_NAMESPACE_NAME);
+  }
+
+  /**
+   * Default and only implementation of {@link ErrorType}
+   */
+  private final static class ErrorTypeImplementation implements ErrorType {
```
Comment (szaffarano) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/error/ErrorTypeBuilder.java:94]: `private static final class`?

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ErrorMappingSourceTypeReferencesExist.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import org.mule.runtime.extension.api.error.ErrorMapping;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/**
+ * Referenced error types do exist in the context of the artifact
+ */
+public class ErrorMappingSourceTypeReferencesExist extends AbstractErrorTypesValidation {
```
Comment (szaffarano) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ErrorMappingSourceTypeReferencesExist.java:30]: This class is quite similar to `ErrorHandlerOnErrorTypeExists`.  Why not extract the common code to a superclass?

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ErrorMappingTargetTypeReferencesExist.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * Referenced error types do exist in the context of the artifact
+ */
+public class ErrorMappingTargetTypeReferencesExist extends AbstractErrorTypesValidation {
```
Comment (szaffarano) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ErrorMappingTargetTypeReferencesExist.java:33]: Same comment that with `ErrorHandlerOnErrorTypeExists`

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/error/MuleCoreErrorTypeRepositoryTestCase.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import org.mule.runtime.api.message.ErrorType;
+
+import java.util.Optional;
+
+import org.junit.Test;
+
+public class MuleCoreErrorTypeRepositoryTestCase {
+
+  @Test
+  public void lookupsAvailableErrorType() {
```
Comment (szaffarano) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/error/MuleCoreErrorTypeRepositoryTestCase.java:25]: it should be named `lookupAvailableErrorType`? (without `s`)

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java (hunk 1 above)
Comment (elrodro83) [reply modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ApplicationPolicyDeploymentTestCase.java:633]: that's what MULE-19203 is about

**Review by szaffarano — APPROVED** [review]

**Pull request merged**